[Docs][KubeRay] Add doc for kubectl plugin (#47851)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/kubectl-plugin.md

(kubectl-plugin)=

# Use Kubectl Plugin (Alpha Feature)

Today it is incredibly challenging for data scientists and AI researchers unfamiliar with Kubernetes to start using Ray on Kubernetes. However, running Ray on Kubernetes is advantageous or necessary in certain environments.

Therefore, starting from KubeRay v1.2.2, we have introduced a `kubectl ray` plugin to simplify the process of running Ray on Kubernetes. The plugin is designed to help users who are not familiar with Kubernetes to run Ray on Kubernetes with ease.

## Installation

> Please follow the instructions in the [KubeRay Kubectl Plugin](https://github.com/ray-project/kuberay/tree/master/kubectl-plugin) to install the plugin.

Reviewer: We bundle the plugin binary into the KubeRay release now right? can we just download the bianry from the release?

Reviewer: Something like:
```
curl -LO https://github.com/ray-project/kuberay/releases/download/v1.2.2/kubectl-ray_v1.2.2_linux_amd64
cp kubectl-ray_v1.2.2_linux_amd64 /usr/local/bin/kubectl-ray 
```

Author: Issue created: https://github.com/ray-project/kuberay/issues/2449

Reviewer: Thanks I think we can directly add this part into this page too:
```
curl -LO https://github.com/ray-project/kuberay/releases/download/v1.2.2/kubectl-ray_v1.2.2_linux_amd64.tar.gz
tar -xvf kubectl-ray_v1.2.2_linux_amd64.tar.gz
cp kubectl-ray ~/.local/bin
```

Author: Done

Downloading file ./log_monitor.log for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./dashboard_agent.log for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./agent-424238335.out for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./agent-424238335.err for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./runtime_env_agent.out for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./runtime_env_agent.err for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./runtime_env_agent.log for Ray Head ray-cluster-kuberay-head-qkrv8
Downloading file ./nsight/ for Ray Head ray-cluster-kuberay-head-qkrv8
```

## Submit a Ray job

This is a wrapper around the `ray job submit` command. The plugin will automatically forward the ports to the Ray cluster and submit the job.

Assume that under the current directory, you have a file named `task.py`.

```python
import ray
ray.init(address="auto")

@ray.remote
def f(x):
    return x * x

futures = [f.remote(i) for i in range(4)]
print(ray.get(futures)) # [0, 1, 4, 9]
```

And a file named `ray-job.usermode.yaml`

```yaml
apiVersion: ray.io/v1
kind: RayJob
metadata:
  name: rayjob-usermode
spec:

>   submissionMode: UserMode

Reviewer: Given we're about to change this for v1.3, should we remove any examples of `kubctl ray job submit`? I'm worried users will try with UserMode with 1.2 and then we'll break them in 1.3

Author: Done in https://github.com/ray-project/ray/pull/47851/commits/c87ec4abd47f7fd52c43b688556d854f3664e886. `kubectl ray job submit` example is completely removed.

(kubectl-plugin)=

# Use Kubectl Plugin (alpha)

Starting from KubeRay v1.2.2, you can use the `kubectl ray` plugin to simplify common workflows when deploying Ray on Kubernetes. If you aren't familiar with Kubernetes, this plugin simplifies running Ray on Kubernetes.

## Installation

> See [KubeRay Kubectl Plugin](https://github.com/ray-project/kuberay/tree/master/kubectl-plugin) to install the plugin.

Reviewer 2: ```suggestion
See [KubeRay kubectl Plugin](https://github.com/ray-project/kuberay/tree/master/kubectl-plugin) to install the plugin.
```

Install the Kuberay kubectl plugin using one of the following methods:

> 

Reviewer 2: ```suggestion

- Install using Krew `kubectl` (recommended)
- Download from GitHub releases

```

Reviewer 2: I meant to add this comment. If you can include this, that would be nice. Thanks!

(kubectl-plugin)=

> # Use Kubectl Plugin (Alpha Feature)

Reviewer 2: ```suggestion
# Use Kubectl plugin (alpha)
```

Reviewer 2: Sentence case for titles.

> Starting from KubeRay v1.2.2, we have introduced a `kubectl ray` plugin to simplify common workflows when deploying Ray on Kubernetes. The plugin is designed to help users who are not familiar with Kubernetes to run Ray on Kubernetes with ease.

Reviewer 2: ```suggestion
Starting from KubeRay v1.2.2, you can use the `kubectl ray` plugin to simplify common workflows when deploying Ray on Kubernetes. If you aren't familiar with Kubernetes, this plugin simplifies running Ray on Kubernetes.
```

## Installation

> Please follow the instructions in the [KubeRay Kubectl Plugin](https://github.com/ray-project/kuberay/tree/master/kubectl-plugin) to install the plugin.

Reviewer 2: ```suggestion
See [KubeRay Kubectl Plugin](https://github.com/ray-project/kuberay/tree/master/kubectl-plugin) to install the plugin.
```

> You can install the Kuberay kubectl plugin using one of the following methods:

Reviewer 2: ```suggestion
Install the Kuberay kubectl plugin using one of the following methods:

-  Use the Krew kubectl plugin manager (recommended)
-  Download from GitHub releases

```

> ### Install using Krew kubectl plugin manager (Recommended)

Reviewer 2: ```suggestion
### Install using the Krew kubectl plugin manager (recommended)
```

1. Install [Krew](https://krew.sigs.k8s.io/docs/user-guide/setup/install/).
2. Download the plugin list by running `kubectl krew update`.
3. Install the plugin by running `kubectl krew install ray`.
4. Run `kubectl ray --help` to verify the installation.

### Download from GitHub releases

Go to the [releases page](https://github.com/ray-project/kuberay/releases) and download the binary for your platform.

For example, to install kubectl plugin version 1.2.2 on Linux amd64:

```bash
curl -LO https://github.com/ray-project/kuberay/releases/download/v1.2.2/kubectl-ray_v1.2.2_linux_amd64.tar.gz
tar -xvf kubectl-ray_v1.2.2_linux_amd64.tar.gz
cp kubectl-ray ~/.local/bin
```

Replace `~/.local/bin` with the directory in your `PATH`.

## Usage

After installing the plugin, you can use `kubectl ray --help` to see the available commands and options.

## Example

> Here assume you have a Ray cluster running on Kubernetes. Please see {ref}`RayCluster Quickstart <kuberay-raycluster-quickstart>` if you don't have a Ray cluster running on Kubernetes.

Reviewer 2: ```suggestion
This example assumes you have a Ray cluster running on Kubernetes. See {ref}`RayCluster Quickstart <kuberay-raycluster-quickstart>` if you don't have a Ray cluster running on Kubernetes.
```

### Get all Ray clusters

```shell

> $ kubectl ray cluster get

Reviewer 3: Maybe remove this section for now. If I remember correctly, we decided to change to `kubectl ray get [cluster | job | service]` to make the syntax closer to `kubectl get [pods, service, ...]`. We can add it back after we change the syntax.

Reviewer 3: cc @andrewsykim

## Usage

After installing the plugin, you can use `kubectl ray --help` to see the available commands and options.

## Example

Here assume you have a Ray cluster running on Kubernetes. Please see {ref}`RayCluster Quickstart <kuberay-raycluster-quickstart>` if you don't have a Ray cluster running on Kubernetes.

### Get all Ray clusters

```shell
$ kubectl ray cluster get

NAME                  NAMESPACE   DESIRED WORKERS   AVAILABLE WORKERS   CPUS   GPUS   TPUS   MEMORY   AGE
ray-cluster-kuberay   default     1                                     2      0      0      3G       24s
```

### Forward local ports to Ray cluster

```shell
$ kubectl ray session ray-cluster-kuberay

Forwarding ports to service ray-cluster-kuberay-head-svc
Ray Dashboard: http://localhost:8265
Ray Interactive Client: http://localhost:10001

Forwarding from 127.0.0.1:8265 -> 8265
Forwarding from [::1]:8265 -> 8265
Forwarding from 127.0.0.1:10001 -> 10001
Forwarding from [::1]:10001 -> 10001
```

### Get Ray cluster logs

```shell
$ kubectl ray log ray-cluster-kuberay

No output directory specified, creating dir under current directory using cluster name.

> Downloading file ./ for Ray Head ray-cluster-kuberay-head-qkrv8

Reviewer 4: Minor change with https://github.com/ray-project/kuberay/pull/2442. The output is now just the ray node name. Ex:
```
$ kubectl ray logs rayjob-sample-raycluster-kfhl6
No output directory specified, creating dir under current directory using cluster name.
Command set to retrieve both head and worker node logs.
Downloading log for Ray Node rayjob-sample-raycluster-kfhl6-head-87xpb
Downloading log for Ray Node rayjob-sample-raycluster-kfhl6-small-group-worker-54qfm
```

Author: Addressed in https://github.com/ray-project/ray/pull/47851/commits/91a1ee2560411a61f0ad6219dd3bb6480d7df085

(kubectl-plugin)=

# Use Kubectl Plugin (Alpha Feature)

> Today it is incredibly challenging for data scientists and AI researchers unfamiliar with Kubernetes to start using Ray on Kubernetes. However, running Ray on Kubernetes is advantageous or necessary in certain environments.

Reviewer: I think we can remove this sentence. This was important context for why we built it, but we don't need to include it in the guide. 
```
Starting from KubeRay v1.2.2, we have introduced a `kubectl ray` plugin to simplify common workflows when deploying Ray on Kubernetes. The plugin is designed to help users who are not familiar with Kubernetes to run Ray on Kubernetes with ease.
```

Author: Done in https://github.com/ray-project/ray/pull/47851/commits/e802c4ba0bb1ec5b5624ef664180b73e5e72426b
